### PR TITLE
Fix template.yaml for compressed bitmanip instructions, Remove print statements from generator.py, Add corner case for division operations, Fix OpComb for amocas.d_32 and amocas.q

### PR DESCRIPTION
**Pull request by Sowmya3062**
1. Running the dev branch gave a yaml scanner error- fixed indentations in template.yaml
2. Removed random print statements from generator.py
3. Solution for the corner case of division was not found- replaced 0x8000000000000000 with  -(2**(XLEN-1)) in dataset.cgf and also added it to rs1_val_data in imc.yaml
4. Corner case of division was added to all rfmt_op_comb which was tested for other instructions apart from division- removed the case from rfmt_op_comb, added a new covergroup with this case to test div and rem instructions only
5. Solution for Op Combination could not be found for amocas.d_32 and amocas.q instructions in the Zacas extension- replaced rv32rv64pair_regs with all_regs for rs2_op_data and rd_op_data in template.yaml
6. rv32ic.cgf did not write tests for c.jr and c.jalr instructions- added op_comb and val_comb for these instructions in rv32ic.cgf
7. rv32i_zcb.cgf did not write tests for the cnot instruction- added op_comb for cnot in rv32i_zcb.cgf
8. rs1_val was not defined for [c.li](http://c.li/), c.lwsp, c.jal, c.jr, c,jalr- added rs1_val_data for these instructions in imc.yaml

### CODE REVIEW

**Review by allenjbaum — COMMENTED** [review]
It helps reviewers when you not only tell them that you fixed something, but say why it needed fixing

File: riscv_ctg/data/imc.yaml [hunk 1]
```diff
@@ -1846,11 +1849,13 @@ c.jr:
...
   formattype: 'crformat'
+  rs1_val_data: 'gen_sign_dataset(xlen)+ gen_sp_dataset(xlen)'
+  rs2_val_data: 'gen_sign_dataset(xlen)+ gen_sp_dataset(xlen)'
   template: |-
 
     // $comment
-    // opcode: c.jr; op1:$rs1
-    TEST_CJR_OP($testreg, $rs1, $swreg, $offset)
+    // opcode: c.jr; op1:$rs1; op2:$rs2; op1val:$rs1_val; op2val:$rs2_val
+    TEST_CJR_OP($testreg, $rs1, $rs2, $swreg, $offset)
```
Comment (pawks) [inline riscv_ctg/data/imc.yaml:1858]: This change does not look correct. The template does not match the macro definition in the `arch_test.h` file.

File: riscv_ctg/data/imc.yaml [hunk 2]
```diff
@@ -1862,9 +1867,11 @@ c.jalr:
...
     - IC
   formattype: 'crformat'
+  rs1_val_data: 'gen_sign_dataset(xlen)+ gen_sp_dataset(xlen)'
+  rs2_val_data: 'gen_sign_dataset(xlen)+ gen_sp_dataset(xlen)'
   template: |-
 
     // $comment
-    // opcode:c.jalr; op1:$rs1
-    TEST_CJALR_OP($testreg, $rs1, $swreg, $offset)
+    // opcode: c.jr; op1:$rs1; op2:$rs2; op1val:$rs1_val; op2val:$rs2_val
```
Comment (pawks) [inline riscv_ctg/data/imc.yaml:1875]: These changes do not look correct. Were the macro files changed? If so, they aren't included in the PR.

**Review by pawks — CHANGES_REQUESTED** [review]
The PR looks incomplete.

File: riscv_ctg/data/imc.yaml (hunk 1 above)
Comment (Sowmya3062) [reply riscv_ctg/data/imc.yaml]: Reverted the change since the template did not require the change.

File: riscv_ctg/data/imc.yaml (hunk 2 above)
Comment (Sowmya3062) [reply riscv_ctg/data/imc.yaml]: Reverted the change since the template did not require the change.

**Comment by Sowmya3062** [comment]
> It helps reviewers when you not only tell them that you fixed something, but say why it needed fixing

Sure.

**Comment by Sowmya3062** [comment]
rv64ic.cgf did not write tests for c.jr and c.jalr instructions- added op_comb and val_comb for these instructions in rv64ic.cgf

**Comment by Sowmya3062** [comment]
9. rv64ic.cgf did not write tests for c.jr and c.jalr instructions- added op_comb and val_comb for these instructions in rv64ic.cgf
10. rv64i_zcb.cgf did not write tests for the cnot instruction- added op_comb for cnot in rv64i_zcb.cgf
11. xlenlim is not defined in rv32i_b.cgf, rv64i_b.cgf- commented xlenlim out of val_comb
12. the formats of leading_ones, leading_zeros,  trailing_ones, trailing_zeros in the val_comb of clz, ctz, cpop were incorrect in rv32i_b.cgf- changed the format
13. the formats of leading_ones, leading_zeros,  trailing_ones, trailing_zeros in the val_comb of clz, clzw, ctz, ctzw, cpop, cpopw were incorrect in rv32i_b.cgf- changed the format
14. rs1_val_data and rs2_val_data were not defined for rolw in template.yaml- added rs1_val_data and rs2_val_data
15. imm_val_data was not defined for bclri, bexti, binvi, bseti in template.yaml- added imm_val_data for these instructions

**Comment by Sowmya3062** [comment]
9. rv64ic.cgf did not write tests for c.jr and c.jalr instructions- added op_comb and val_comb for these instructions in rv64ic.cgf
10. rv64i_zcb.cgf did not write tests for the cnot instruction- added op_comb for cnot in rv64i_zcb.cgf
11. xlenlim is not defined in rv32i_b.cgf, rv64i_b.cgf- commented xlenlim out of val_comb
12. The formats of leading_ones, leading_zeros, trailing_ones, trailing_zeros in the val_comb of clz, ctz, cpop were incorrect in rv32i_b.cgf and rv32e_b.cgf- changed the format
13. The formats of leading_ones, leading_zeros, trailing_ones, trailing_zeros in the val_comb of clz, clzw, ctz, ctzw, cpop, cpopw were incorrect in rv32i_b.cgf- changed the format
14. rs1_val_data and rs2_val_data were not defined for rolw in template.yaml- added rs1_val_data and rs2_val_data
15. imm_val_data was not defined for bclri, bexti, binvi, bseti in template.yaml- added imm_val_data for these instructions

**Comment by neelgala** [comment]
@pawks the only thing still not working from sowmya's branch is rv32ip.cgf - needs much more deep dive to fix it and will come in later. If you agree can we merge this to dev - sowmya can then starting validating the other PRs and getting them merged. Primary attempt is to reduce the backlog on PRs asap.

There is also this new parameter called xlenlim used in a few bitmanip op - the definition of which doesn't exist anywhere (neither here nor in isac). @Sowmya3062 please file an issue for this bug on the repo.

@allenjbaum almost everything was broken (even the basic rv32i.cgf wasn't working) on the dev branch and hence this PR. The PR is supposed get a clean slate to start merging other PRs with more confidence.

**Comment by ved-rivos** [comment]
> Solution for Op Combination could not be found for amocas.d_32 and amocas.q instructions in the Zacas extension- replaced rv32rv64pair_regs with all_regs for rs2_op_data and rd_op_data in template.yaml

@Sowmya3062 - The ISA requires pair regs for amocas.d_32 and amocas.q - so replacing with all_regs is wrong.

**Comment by Sowmya3062** [comment]
> > Solution for Op Combination could not be found for amocas.d_32 and amocas.q instructions in the Zacas extension- replaced rv32rv64pair_regs with all_regs for rs2_op_data and rd_op_data in template.yaml
> 
> @Sowmya3062 - The ISA requires pair regs for amocas.d_32 and amocas.q - so replacing with all_regs is wrong.

Reverted the changes. Instructions amocas.w, amocas.d_32 and amocas.d_64 now use register pairs for rs2 and rd.

**Review by pawks — APPROVED** [review]

**Review by pawks — CHANGES_REQUESTED** [review]
This PR seems to add riscv-isac as a submodule in riscv-ctg. I dont think that this should be done.

**Comment by Sowmya3062** [comment]
> This PR seems to add riscv-isac as a submodule in riscv-ctg. I dont think that this should be done.

Deleted riscv-isac from riscv-ctg

**Review by pawks — APPROVED** [review]
